### ISSUE TITLE
Indentation: Object initialization with double-curlies is not allowed

### ISSUE THREAD

**Issue by JoshMcCullough** [checkstyle/checkstyle#11307]
Docs: https://checkstyle.sourceforge.io/config_misc.html#Indentation

```bash
/var/tmp $ javac Test.java
# compiles fine; no output

/var/tmp $ cat config.xml

<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd"> 

<module name="Checker">
    <module name="TreeWalker">
        <module name="Indentation"/>
    </module>
</module>   

/var/tmp $ cat Test.java

public class Test
{
    private final MyThing thing;

    public Test()
    {
        this.thing = new MyThing()
        {{
            setName("something");
        }};
    }

    private static class MyThing
    {
        private String name;

        public String getName()
        {
            return name;
        }

        public void setName(String name)
        {
            this.name = name;
        }
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.0-all.jar -c config.xml Test.java

Starting audit...
[ERROR] /home/josh/Desktop/checkstyle-test/api/src/main/java/Test.java:9:13:
     'block' child has incorrect indentation level 12, expected level should be 16. [Indentation]
[ERROR] /home/josh/Desktop/checkstyle-test/api/src/main/java/Test.java:10:9:
     'block rcurly' has incorrect indentation level 8, expected level should be 12. [Indentation]
Audit done.
Checkstyle ends with 2 errors.
```

---

object initialization should be supported as it is just static block in anonymous class declaration, but is being flagged by the `Indentation` check. Attention that there is no violations on methods/class blocks that starts at new line with same indentation.

According to the indentation docs, perhaps double-curly object initialization is just not supported unless it's formatted as follows (which of course works):

```java
this.thing = new MyThing()
{
    {
        setName("something");
    }
};
```

**Comment by nmancus1** [comment]
@JoshMcCullough 

> According to https://github.com/checkstyle/checkstyle/issues/2838, object initialization should be supported

The issue that you have linked to does not relate to `Indentation` check; each Checkstyle check is modular and does not depend on other checks.

Please update your issue description to follow issue template, i.e. read check documentation and add a link to documentation, and refine your configuration to only contain the related check (`Indentation`).

**Comment by JoshMcCullough** [comment]
@nmancus1 Updated as requested. According to the indentation docs, perhaps double-curly object initialization is just not supported unless it's formatted as follows (which of course works):

```java
this.thing = new MyThing()
{
    {
        setName("something");
    }
};
```

**Comment by nmancus1** [comment]
I am not sure that we should support this. "double-brace initialization" isn't a particular syntax, rather it is just a convenient way to initialize an anonymous inner class. Consider:

```
    final List<String> someList = new ArrayList<>()
    {

        // some instance initializer
        {
            this.add("some string");
        } 

        @Override
        public int size() {
            return super.size();
        }
    };

```

This is the same exact syntax. Would we treat this differently?

As an alternative, you could use `SuppressionXpathSingleFilter` to suppress warnings within anonymous class initializer blocks like so:

```
➜  src javac Test.java

➜  src cat Test.java  
import java.util.ArrayList;
import java.util.List;

public class Test
{
    private final MyThing thing;

    public Test()
    {
        this.thing = new MyThing()
        {{
            setName("something");
        }};
    }

    private static class MyThing
    {
        private String name;

        public String getName()
        {
            return name;
        }

        public void setName(String name)
        {
            this.name = name;
        }
    }
}

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="Indentation"/>
        <module name="SuppressionXpathSingleFilter">
            <property name="checks" value="Indentation"/>
            <property name="query" value="//LITERAL_NEW//INSTANCE_INIT/SLIST//*"/>
        </module>
    </module>
</module>

➜  src java -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
Audit done.


```

**Comment by JoshMcCullough** [comment]
@nmancus1 given that this "double-brace initialization" is specifically mentioned/supported in the `WhitespaceAround` module (https://github.com/checkstyle/checkstyle/issues/2838), it seemed like it should be allowed here, so things are consistent.

>This check does not flag as violation double brace initialization like:

```java
new Properties() {{
    setProperty("key", "value");
}};
```

But I'll give your suggestion a try, thanks a lot!

**Comment by JoshMcCullough** [comment]
Note: it works, so we'll go with it for now but I'd love to see official support!

**Comment by JoshMcCullough** [comment]
@romani Can you elaborate on how you ran the test in the OP? I just ran it again to make sure I wasn't crazy. Here are the commands I used and the output for 9.3 and 10.

```bash
> cat <<EOT > Test.java                                               
public class Test      
{                                                        
    private final MyThing thing;                                

    public Test()      
    {                         
        this.thing = new MyThing()  
        {{   
            setName("something");
        }};
    }

    private static class MyThing
    {
        private String name;

        public String getName()
        {
            return name;
        }

        public void setName(String name)
        {
            this.name = name;
        }
    }
}

EOT
> cat <<EOT > config.xml                                              
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd"> 

<module name="Checker">
    <module name="TreeWalker">
        <module name="Indentation"/>
    </module>
</module>    

EOT
> javac Test.java                                                     
> RUN_LOCALE="-Duser.language=en -Duser.country=US"
> java $RUN_LOCALE -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/josh/Downloads/Test.java:9:13: 'block' child has incorrect indentation level 12, expected level should be 16. [Indentation]
[ERROR] /home/josh/Downloads/Test.java:10:9: 'block rcurly' has incorrect indentation level 8, expected level should be 12. [Indentation]
Audit done.
Checkstyle ends with 2 errors.
 ✘> java $RUN_LOCALE -jar checkstyle-10.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/josh/Downloads/Test.java:9:13: 'block' child has incorrect indentation level 12, expected level should be 16. [Indentation]
[ERROR] /home/josh/Downloads/Test.java:10:9: 'block rcurly' has incorrect indentation level 8, expected level should be 12. [Indentation]
Audit done.
Checkstyle ends with 2 errors.
```

**Comment by romani** [comment]
> ran the test in the OP

what is OP ?
you did good execution and sharing all details.

**Comment by JoshMcCullough** [comment]
@romani thanks. And "OP" is "original post" e.g. my first post at the top. @nick-mancuso's solution works around the problem. I'm wondering if originally you tested using the [example in his comment](https://github.com/checkstyle/checkstyle/issues/11307#issuecomment-1039804246) rather than my original example?

**Comment by romani** [comment]
I usually do not run, until issue reported properly.

Please answer my question about
>This check does not flag as violation double brace initialization like

**Comment by JoshMcCullough** [comment]
Sorry but I'm not quite following you. My original post has a working (failing) scenario, the same as my [updated comment](https://github.com/checkstyle/checkstyle/issues/11307#issuecomment-1058306811).

Regarding that quote...

> This check does not flag as violation double brace initialization like

That is a direct quote from the docs for `WhiteSpaceAround`: https://checkstyle.sourceforge.io/config_whitespace.html#WhitespaceAround

So, with double-brace initializaion...

```java
this.thing = new MyThing() {{
    setName("something");
}};
```

The `WhiteSpaceAround` check will pass but the `Indentation` check will fail. I suggest they should both come up with the same resolution (allow the double-brace initialization syntax).

**Comment by JoshMcCullough** [comment]
@romani I updated my original post, adding a bit more clarity and a more complete example.

**Comment by romani** [comment]
I am not sure how to support it in current design of this Check, but issue is valid, let's keep it, one day we will do big redesign, so we need to know this request.
